third_party/validator/constants.go

```diff
@@ -1,10 +1,16 @@
 package google
 
 import (
+	"errors"
+
 	"github.com/hashicorp/terraform/helper/mutexkv"
 )
 
+// ErrNoConversion can be returned if a conversion is unable to be returned
+// because of the current state of the system.
+// Example: The conversion requires that the resource has already been created
+// and is now being updated).
+var ErrNoConversion = errors.New("no conversion")
```

Comment:
Downstream we are using http://github.com/pkg/errors. The pattern we use is: for actionable errors, define a well-known error using `errors.New` (which produces a unique pointer value to compare on, i.e. `errors.New("a") != errors.New("a")`, as you probably know), wrap the error with specific details using `.Wrap` as it gets handed up the stack, and inspect the root cause at the error-handling location using `.Cause`.

Other approaches that I am aware of:
2. Assertions surrounding interfaces (https://dave.cheney.net/tag/error-handling)
3. Assertions surrounding concrete types (`type MyError struct { ...`)

Unfortunately all of these approaches are no longer easily inspected downstream once `fmt.Errorf` comes into the picture somewhere in the stack, so no matter the approach we take here, we should probably scrub all calls to `fmt.Errorf` as a means of adding context and handing up the stack.

It sounds like you are in favor of 3 as a strategy for the conversions library?

Reply:
For reference: https://github.com/GoogleCloudPlatform/terraform-validator/pull/8/files#diff-747eeb1f3694357a501140221061b33eR166

Reply:
Ah I didn't realize you were using pkg/errors downstream. This more sense when I realize this is a sentinel error and not just a string :) Definitely agree that `errors.WrapF` is far superior to `fmt.errorF`.

In isolation I'm on the fence between using sentinel errors vs error type assertion. Our cloud go clients use the error types pretty heavily and that means in the provider code we do a fair amount of [the latter](https://github.com/terraform-providers/terraform-provider-google/blob/master/google/utils.go#L353) so I have a weak preference for that.

Reply:
I agree with the custom error type approach if you have well known metadata that corresponds to the error. HTTP/gRPC status codes are a great example there. I prefer the sentinel error in cases where context is better added as a freeformed string (via `.Wrap`). Let me know which direction you want to go.

Reply:
IMO it depends on the behavior that you are looking for downstream rather than the context getting added. I think both approaches get a lot of value by adding context with `Wrap`. If there are categories of errors that require different behavior (eg: retriable vs non-retriable) then using custom types makes more sense. I'm not sure how the downstream consumer will be handling 'No Conversion' errors but I had imagined it to be `ignorable` or something like that.

Either way I think you can make the call on which approach.